Model/src/main/java/org/apidb/apicommon/model/report/bed/feature/GeneComponentsFeatureProvider.java

```diff
@@ -0,0 +1,124 @@
+  private final Set<GeneComponent> _requestedComponents;
+
+  public GeneComponentsFeatureProvider(JSONObject config){
+    _requestedDeflineFields = new RequestedDeflineFields(config);
+    _requestedComponents = config.getJSONArray("geneComponents").toList().stream().map(o -> GeneComponent.valueOf(o.toString())).collect(Collectors.toSet());
+
+  }
+  @Override
+  public String getRequiredRecordClassFullName() {
+    return TranscriptUtil.GENE_RECORDCLASS;
+  }
+
+  @Override
+  public String[] getRequiredAttributeNames() {
+    return new String[] { ATTR_ORGANISM } ;
+  }
+
+  @Override
+  public String[] getRequiredTableNames() {
+    return new String[] { TABLE_GENE_MODEL_DUMP };
+  }
+
+  @Override
+  public List<List<String>> getRecordAsBedFields(RecordInstance record) throws WdkModelException {
+    try {
+      List<List<String>> result = new ArrayList<>();
+      Map<String, List<Integer>> startsByT = new HashMap<>();
+      Map<String, List<Integer>> endsByT = new HashMap<>();
+      TableValue geneModelDumpRows = record.getTableValue(TABLE_GENE_MODEL_DUMP);
+
+      for (Map<String, AttributeValue> geneModelDumpRow : geneModelDumpRows) {
+        String sourceId = geneModelDumpRow.get("source_id").toString();
+        String chrom = geneModelDumpRow.get("sequence_id").toString();
+        GeneComponent geneComponent = GeneComponent.valueOf(geneModelDumpRow.get("type").toString().toLowerCase());
+        Integer segmentStart = Integer.valueOf(geneModelDumpRow.get("gm_start").toString());
+        Integer segmentEnd = Integer.valueOf(geneModelDumpRow.get("gm_end").toString());
+
+        StrandDirection strand = StrandDirection.fromSign(geneModelDumpRow.get("strand").toString());
+
+        if(_requestedComponents.contains(geneComponent)){
+          DeflineBuilder defline = new DeflineBuilder(sourceId + "::" + geneComponent.toString());
+          if(_requestedDeflineFields.contains("organism")){
+            defline.appendRecordAttribute(record, ATTR_ORGANISM);
+          }
+          if(_requestedDeflineFields.contains("description")){
+            defline.appendValue(prettyString(geneComponent) + " sequence");
+          }
+          if(_requestedDeflineFields.contains("position")){
+            defline.appendPosition(chrom, segmentStart, segmentEnd, strand);
+          }
+          if(_requestedDeflineFields.contains("ui_choice")){
+            defline.appendValue("Gene components: " + _requestedComponents.stream().map(x -> prettyString(x)).collect(Collectors.joining(", ")));
+          }
+          if(_requestedDeflineFields.contains("segment_length")){
+            defline.appendSegmentLength(segmentStart, segmentEnd);
+          }
+          result.add(BedLine.bed6(chrom, segmentStart, segmentEnd, defline, strand));
+        }
+      }
+      return result;
+    }
+    catch (WdkUserException e){
+      throw new WdkModelException(e);
+    }
+  }
+
+  private String prettyString(GeneComponent geneComponent){
```

Comment:
Consider making this a private field of the enum.  Then you don't have to worry about exceptions and if anyone ever wants to add an enum value, they know they also need to add a pretty string.